File: app/src/main/java/no/joharei/flixr/mainpage/MainPresenter.java
package no.joharei.flixr.mainpage;

import android.util.Log;

import javax.inject.Inject;

import no.joharei.flixr.MainApplication;
import no.joharei.flixr.tools.RxAssist;

public class MainPresenter {
    private static final String TAG = MainPresenter.class.getSimpleName();
    @Inject
    MainApi mainApi;
    private MainView view;

    void attachView(MainView view) {
        this.view = view;
        MainApplication.component(view.getContext()).inject(this);
    }

    void fetchMyPhotosets() {
        mainApi.getMyPhotosets()
                .compose(RxAssist.applyDefaultSchedulers())
                .subscribe(photosets -> {
                    view.showMyPhotosets(photosets.getPhotosets());
                }, throwable -> {
                    Log.e(TAG, "Failed fetching my photosets", throwable);
                    // TODO
                });
    }

    void fetchMyContacts() {
        mainApi.getContacts()
                .compose(RxAssist.applyDefaultSchedulers())
                .subscribe(contacts -> {
                    view.showMyContacts(contacts.getContacts());
                }, throwable -> {
                    Log.e(TAG, "Failed fetching my contacts", throwable);
                    // TODO
                });
    }
}

File: app/src/main/java/no/joharei/flixr/mainpage/MainPresenter.kt
package no.joharei.flixr.mainpage

import io.reactivex.disposables.CompositeDisposable
import no.joharei.flixr.MainApplication
import no.joharei.flixr.tools.applyDefaultSchedulers
import timber.log.Timber
import javax.inject.Inject

class MainPresenter {
    @Inject
    lateinit var mainApi: MainApi
    lateinit var view: MainView
    private val compositeSubscription = CompositeDisposable()

    fun attachView(view: MainView) {
        this.view = view
        MainApplication.component.inject(this)
    }

    fun fetchMyPhotosets() {
        val photosetsSub = mainApi.fetchMyPhotosets()
                .applyDefaultSchedulers()
                .subscribe(
                        { photosets -> view.showMyPhotosets(photosets.photosets) },
                        { throwable ->
                            Timber.e(throwable, "Failed fetching my photosets")
                            mainApi.clearCache()
                            // TODO
                        })
        compositeSubscription.add(photosetsSub)
    }

    fun fetchMyContacts() {
        val contactsSub = mainApi.fetchContacts()
                .applyDefaultSchedulers()
                .subscribe(
                        { contacts -> view.showMyContacts(contacts.contacts) },
                        { throwable ->
                            Timber.e(throwable, "Failed fetching my contacts")
                            mainApi.clearCache()
                            // TODO
                        })
        compositeSubscription.add(contactsSub)
    }

    fun stop() {
        compositeSubscription.clear()
    }
}
